test.sh

```diff
@@ -1,40 +1,72 @@
-#!/bin/bash
+#!/usr/bin/env bash
+
+# -e Stops execution in the instance of a command or pipeline error
+# -u Treat unset variables as an error and exit immediately
+set -eu
+
 # Run all tests and coverage checks. Called from Travis automatically, also
 # suitable to run manually. See list of prerequisite packages in .travis.yml
 if type realpath >/dev/null 2>&1 ; then
-  cd $(realpath $(dirname $0))
+  cd "$(realpath -- $(dirname -- "$0"))"
 fi
 
-set -exo pipefail
-TRAVIS=${TRAVIS:-false}
+#
+# Defaults
+#
+STATUS="FAILURE"
+if [ -z "${RUN+x}" ]; then RUN=(); fi
+if [ -z "${UNIT_FILTER+x}" ]; then UNIT_FILTER=(); fi
+if [ -z "${INT_FILTER+x}" ]; then INT_FILTER=(); fi
+TRAVIS="${TRAVIS:-false}"
+export BOULDER_CONFIG_DIR="${BOULDER_CONFIG_DIR:-"test/config"}"
 
-# The list of segments to run. To run only some of these segments, pre-set the
-# RUN variable with the ones you want (see .travis.yml for an example).
-# Order doesn't matter. Note: gomod-vendor is specifically left out of the
-# defaults, because we don't want to run it locally (it could delete local
-# state) We also omit coverage by default on local runs because it generates
-# artifacts on disk that aren't needed.
-RUN=${RUN:-lints unit integration}
+#
+# Print Functions
+#
+function print_outcome() {
+  if [ "$STATUS" == SUCCESS ]
+  then
+    echo -e "\e[32m"$STATUS"\e[0m"
+  else
+    echo -e "\e[31m"$STATUS"\e[0m"
+  fi
+}
+
+function print_list_of_integration_tests() {
+  for file in ./test/integration/*.go; do
+    [ -e "$file" ] || continue
+    cat "$file" | grep -e '^func Test' | awk '{print $2}' | sed s/\(t//
+  done
+  exit 0
+}
+
+function exit_msg() { echo "$*" >&2; exit 2; }  # complain to STDERR and exit with error
+function check_arg() { if [ -z "$OPTARG" ]; then exit_msg "No arg for --$OPT option, use: -h for help">&2; fi; }
+function trap_outcome_on_exit() { trap "print_outcome" EXIT; }
+function print_usage_exit() { echo "$USAGE"; exit 0; }
+function print_heading { echo; echo -e "\e[34m\e[1m"$1"\e[0m"; }
 
 function run_and_expect_silence() {
   echo "$@"
   result_file=$(mktemp -t bouldertestXXXX)
-  "$@" 2>&1 | tee ${result_file}
+  "$@" 2>&1 | tee "${result_file}"
 
   # Fail if result_file is nonempty.
-  if [ -s ${result_file} ]; then
-    rm ${result_file}
+  if [ -s "${result_file}" ]; then
+    rm "${result_file}"
     exit 1
   fi
-  rm ${result_file}
+  rm "${result_file}"
 }
 
+#
+# Testing Helpers
+#
 function run_unit_tests() {
-  if [ "${TRAVIS}" == "true" ]; then
+  if [ "${TRAVIS}" == true ]; then
```

Comment:
I'd make whether or not to use the `-race` flag a flag at the top level here. Travis can set the flag, users can ignore it. Explicit > Implicit when possible.

Reply:
I'm not sure if I fully understand your feedback here. `TRAVIS` as a var is a holdover from the previous version of the script. However, it's explicitly flagged out here on the `getopts` parser:
```
    e | enable-race-detector )       TRAVIS="true" ;;
```

Reply:
Ah! I'd missed that. And as a result of me missing it, I'd instead say that this behavior should be controlled by a variable with a more descriptive/accurate name (i.e. `TRAVIS` might be true if a user sets `-e` on their local workstation, which would be weird).